patch/llpcPatchResourceCollect.cpp

```diff
@@ -3501,6 +3506,250 @@ void PatchResourceCollect::ReassembleOutputExportCalls()
+            func.getName().startswith(LlpcName::InputImportInterpolant))
+        {
+            // This is a generic (possibly interpolated) input. Find its uses in FS.
+            for (User* pUser : func.users())
+            {
+                auto pCall = cast<CallInst>(pUser);
+                if (m_pPipelineShaders->GetShaderStage(pCall->getFunction()) != ShaderStageFragment)
+                {
+                    continue;
+                }
+                // We have a use in FS. See if it needs scalarizing.
+                if (isa<VectorType>(pCall->getType()) || (pCall->getType()->getPrimitiveSizeInBits() == 64))
+                {
+                    fsInputCalls.push_back(pCall);
+                }
+            }
+        }
+        else if (func.getName().startswith(LlpcName::OutputExportGeneric))
+        {
+            // This is a generic output. Find its uses in the final vertex stage.
+            for (User* pUser : func.users())
+            {
+                auto pCall = cast<CallInst>(pUser);
+                if (m_pPipelineShaders->GetShaderStage(pCall->getFunction()) !=
+                      m_pContext->GetPipelineContext()->GetFinalVertexStage())
+                {
+                    continue;
+                }
+                // We have a use the final vertex stage. See if it needs scalarizing. The output value is
+                // always the final argument.
+                Type* pValueTy = pCall->getArgOperand(pCall->getNumArgOperands() - 1)->getType();
+                if (isa<VectorType>(pValueTy) || (pValueTy->getPrimitiveSizeInBits() == 64))
+                {
+                    vsOutputCalls.push_back(pCall);
+                }
+            }
+        }
+    }
+
+    // Scalarize the gathered inputs and outputs.
+    for (CallInst* pCall : fsInputCalls)
+    {
+        ScalarizeInputGeneric(pCall);
+    }
+    for (CallInst* pCall : vsOutputCalls)
+    {
+        ScalarizeOutputGeneric(pCall);
+    }
+}
+
+// =====================================================================================================================
+// Scalarize a generic input.
+// This is known to be an FS generic or interpolant input that is either a vector or 64 bit.
+void PatchResourceCollect::ScalarizeInputGeneric(
+    CallInst* pCall)  // [in] The llpc.input.import.generic or llpc.input.import.interpolant call
+{
+    IRBuilder<> builder(pCall->getContext());
+    builder.SetInsertPoint(pCall);
+
+    // FS:  @llpc.input.import.generic.%Type%(i32 location, i32 elemIdx, i32 interpMode, i32 interpLoc)
+    //      @llpc.input.import.interpolant.%Type%(i32 location, i32 locOffset, i32 elemIdx,
+    //                                            i32 interpMode, <2 x float> | i32 auxInterpValue)
+    SmallVector<Value*, 5> args;
+    for (uint32_t i = 0, e = pCall->getNumArgOperands(); i != e; ++i)
+    {
+        args.push_back(pCall->getArgOperand(i));
+    }
+
+    bool isInterpolant = args.size() != 4;
```

Comment:
This is just a suggestion: args.size() != 4 -> (args.size() != 4)
I find in LLVM coding, the "()" often be omitted in various expression evaluation and rely on the operator priority. I don't recommend this. We ough to use () to make code logic clear as possible as we could.

Reply:
There are some cases where omitting parentheses is confusing, but there is general consensus on which cases those are, and those all generate errors when you use -Wall -Werror. This is not one of those cases, so I think we should not add extra parentheses over what is generally used in LLVM, to avoid clutter.